src/Adapter/Employee/CommandHandler/ChangePasswordHandler.php

```diff
@@ -0,0 +1,176 @@
+    ) {
+        $this->configuration = $configuration;
+        $this->hashing = $hashing;
+        $this->legacyContext = $legacyContext;
+        $this->translator = $translator;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function handle(ChangePasswordCommand $command)
+    {
+        $employee = new Employee();
+
+        // Legacy condition. Employee object is loaded in $employee->getByEmail() call, so it must go first.
+        $employeeFound = $employee->getByEmail($command->getEmail()->getValue())
+            && $employee
+            && $employee->id == $command->getEmployeeId()->getValue()
+        ;
+
+        if (!$employeeFound) {
+            throw new EmployeeNotFoundException(null, 'Employee could not be found by given email.');
+        }
+
+        if (Password::MIN_LENGTH > strlen($command->getNewPlainPassword())) {
+            throw new EmployeeConstraintException(
+                sprintf(
+                    'Password must be not shorter than %d symbols',
+                    Password::MIN_LENGTH
+                ),
+                EmployeeConstraintException::INVALID_PASSWORD
+            );
+        }
+
+        $timeExpression = sprintf(
+            '%s + %s minutes',
+            (string) $employee->last_passwd_gen,
+            (string) $this->configuration->get('PS_PASSWD_TIME_BACK')
+        );
+        $canChangePassword = (strtotime($timeExpression) - time()) <= 0;
+
+        if (!$canChangePassword) {
+            throw new PasswordChangeTooFrequentException('Password has been changed recently, you must wait.');
+        }
+
+        if ($employee->getValidResetPasswordToken() !== $command->getToken()) {
+            throw new ChangePasswordTokenExpiredException('Change password token has expired.');
+        }
+
+        $employee->passwd = $this->hashing->hash($command->getNewPlainPassword());
+        $employee->last_passwd_gen = date('Y-m-d H:i:s', time());
+
+        $params = array(
+            '{email}' => $employee->email,
+            '{lastname}' => $employee->lastname,
+            '{firstname}' => $employee->firstname,
+        );
+
+        $employeeLanguage = new Language((int) $this->legacyContext->getContext()->employee->id_lang);
+
+        $mailWasSent = Mail::send(
+            $employee->id_lang,
+            'password',
+            $this->translator->trans(
+                'Your new password',
+                [],
+                'Emails.Subject',
+                $employeeLanguage->locale
+            ),
+            $params,
+            $employee->email,
+            $employee->firstname . ' ' . $employee->lastname
+        );
+
+        if (!$mailWasSent) {
+            throw new UnableToChangePasswordException('Password change email could not be sent.');
+        }
+
+        // Update employee only if the mail can be sent
+        Shop::setContext(Shop::CONTEXT_SHOP, (int) min($employee->getAssociatedShops()));
+
+        if (!$employee->update()) {
```

Comment:
About that double update call, I think indeed you could call `$employee->removeResetPasswordToken()` before the first update If the update didn't work the token won't be removed anyway
By the way it would seem more logical if the email was sent after the employee is updated, this would avoid send an email if the employee was not updated AND it would allow to update it even though there is an issue on mail sending

Reply:
>About that double update call, I think indeed you could call $employee->removeResetPasswordToken() before the first update

yes, exactly!